base/cvd/cuttlefish/common/libs/utils/files.cpp

```diff
@@ -81,6 +81,30 @@ bool FileExists(const std::string& path, bool follow_symlinks) {
   return (follow_symlinks ? stat : lstat)(path.c_str(), &st) == 0;
 }
 
+Result<dev_t> FileDeviceId(const std::string& path) {
+  struct stat out;
+  CF_EXPECTF(stat(path.c_str(), &out) == 0,
+             "stat() failed trying to retrive device ID information for \"{}\" "
+             "with error: {}",
+             path, strerror(errno));
+  return out.st_dev;
+}
+
+Result<bool> CanHardLink(const std::string& path1, const std::string& path2) {
+  return CF_EXPECT(FileDeviceId(path1)) == CF_EXPECT(FileDeviceId(path2));
+}
+
+Result<std::string> CreateHardLink(const std::string& target,
+                                   const std::string& hardlink) {
+  if (!FileExists(hardlink)) {
+    CF_EXPECTF(link(target.c_str(), hardlink.c_str()) == 0,
+               "link() failed trying to create hardlink from \"{}\" to \"{}\" "
+               "with error: {}",
+               target, hardlink, strerror(errno));
+  }
+  return hardlink;
```

Comment:
Done.

I added logic to check if the files are already hardlinked.  I also added the ability to force an overwriting the existing file.

We cannot simply error if the file already exists if we want to match existing `cvd fetch` behavior, because right now it overwrites existing files.

We also cannot rely on surfacing an error if the file already exists but is not a hardlink of the file in the cache because some files (like [`misc_info.txt`](https://github.com/google/android-cuttlefish/blob/61c7bc37409dbbbf196599c8901d27483fe1f8df/base/cvd/cuttlefish/host/commands/cvd/fetch/fetch_cvd.cc#L463)) have cases where the file is allowed to be missing and anywhere in the download is allowed to fail.  I discovered while testing the logic that verified the existing file was hardlinked that the error would be silently fail and then not get an entry in the `fetcher_config.json`. The overwriting matches the current behavior.